<a href="https://colab.research.google.com/github/valogonor/DS-Unit-2-Sprint-3-Advanced-Regression/blob/master/DS_Unit_2_Sprint_Challenge_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 2 Sprint Challenge 3

## Logistic Regression and Beyond

In this sprint challenge you will fit a logistic regression modeling the probability of an adult having an income above 50K. The dataset is available at UCI:

https://archive.ics.uci.edu/ml/datasets/adult

Your goal is to:

1. Load, validate, and clean/prepare the data.
2. Fit a logistic regression model
3. Answer questions based on the results (as well as a few extra questions about the other modules)

Don't let the perfect be the enemy of the good! Manage your time, and make sure to get to all parts. If you get stuck wrestling with the data, simplify it (if necessary, drop features or rows) so you're able to move on. If you have time at the end, you can go back and try to fix/improve.

### Hints

It has a variety of features - some are continuous, but many are categorical. You may find [pandas.get_dummies](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html) (a method to one-hot encode) helpful!

The features have dramatically different ranges. You may find [sklearn.preprocessing.minmax_scale](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.minmax_scale.html#sklearn.preprocessing.minmax_scale) helpful!

## Part 1 - Load, validate, and prepare data

The data is available at: https://archive.ics.uci.edu/ml/datasets/adult

Load it, name the columns, and make sure that you've loaded the data successfully. Note that missing values for categorical variables can essentially be considered another category ("unknown"), and may not need to be dropped.

You should also prepare the data for logistic regression - one-hot encode categorical features as appropriate.

In [1]:
import pandas as pd
names = ['age', 'workclass', 'fnlwgt', 'education', 'education_num', 'marital_status', 'occupation', 'relationship', 'race', 'sex', 'capital_gain', 'capital_loss',
         'hours_per_week', 'native_country', 'income']
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data', names=names)
df.head()

age          workclass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [2]:
df.isna().sum()

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64

In [3]:
import numpy as np
df = df.replace(' ?', np.nan)
df.isna().sum()

age                  0
workclass         1836
fnlwgt               0
education            0
education_num        0
marital_status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital_gain         0
capital_loss         0
hours_per_week       0
native_country     583
income               0
dtype: int64

In [4]:
df.shape

(32561, 15)

In [8]:
df = df.replace(np.nan, 'Unknown')
df = df.replace(' <=50K', 0)
df = df.replace(' >50K', 1)
df = df.replace(' Female', 0)
df = df.replace(' Male', 1)
df.head()

age          workclass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital_status          occupation    relationship    race  sex  \
0        Never-married        Adm-clerical   Not-in-family   White    1   
1   Married-civ-spouse     Exec-managerial         Husband   White    1   
2             Divorced   Handlers-cleaners   Not-in-family   White    1   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black    1   
4   Married-civ-spouse      Prof-specialty            Wife   Black    0   

   capital_gain  capital_loss  hours_per_week  native_country  income  
0          2174             0              40   United-States       0  
1             0             0              13   United-States       0  
2             0             0              40   United-States       0  
3             0             0              40   United-States       0  
4             0             0              40            Cuba       0

In [9]:
df.describe()

age        fnlwgt  education_num           sex  capital_gain  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679      0.669205   1077.648844   
std       13.640433  1.055500e+05       2.572720      0.470506   7385.292085   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      1.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      1.000000      0.000000   
max       90.000000  1.484705e+06      16.000000      1.000000  99999.000000   

       capital_loss  hours_per_week        income  
count  32561.000000    32561.000000  32561.000000  
mean      87.303830       40.437456      0.240810  
std      402.960219       12.347429      0.427581  
min        0.000000        1.000000      0.000000  
25%        0.000000       40.000000      0.000000  
50%        0.000000       40.000000      0.000000  
75%        0.000000       45.000000      0.000000  
max     4356.000000       99.000000      1.000000

In [10]:
df_num = df._get_numeric_data().columns.tolist()
df_num

['age',
 'fnlwgt',
 'education_num',
 'sex',
 'capital_gain',
 'capital_loss',
 'hours_per_week',
 'income']

In [11]:
df_cat = df.drop(columns=df_num)
df_cat.head()

workclass   education       marital_status          occupation  \
0          State-gov   Bachelors        Never-married        Adm-clerical   
1   Self-emp-not-inc   Bachelors   Married-civ-spouse     Exec-managerial   
2            Private     HS-grad             Divorced   Handlers-cleaners   
3            Private        11th   Married-civ-spouse   Handlers-cleaners   
4            Private   Bachelors   Married-civ-spouse      Prof-specialty   

     relationship    race  native_country  
0   Not-in-family   White   United-States  
1         Husband   White   United-States  
2   Not-in-family   White   United-States  
3         Husband   Black   United-States  
4            Wife   Black            Cuba

In [12]:
df_catd = pd.get_dummies(df_cat)
pd.set_option('max_columns', 100)
df_catd.head()

workclass_ Federal-gov  workclass_ Local-gov  workclass_ Never-worked  \
0                       0                     0                        0   
1                       0                     0                        0   
2                       0                     0                        0   
3                       0                     0                        0   
4                       0                     0                        0   

   workclass_ Private  workclass_ Self-emp-inc  workclass_ Self-emp-not-inc  \
0                   0                        0                            0   
1                   0                        0                            1   
2                   1                        0                            0   
3                   1                        0                            0   
4                   1                        0                            0   

   workclass_ State-gov  workclass_ Without-pay  workclass_Unknown  \
0                     1                       0                  0   
1                     0                       0                  0   
2                     0                       0                  0   
3                     0                       0                  0   
4                     0                       0                  0   

   education_ 10th  education_ 11th  education_ 12th  education_ 1st-4th  \
0                0                0                0                   0   
1                0                0                0                   0   
2                0                0                0                   0   
3                0                1                0                   0   
4                0                0                0                   0   

   education_ 5th-6th  education_ 7th-8th  education_ 9th  \
0                   0                   0               0   
1                   0                   0               0   
2                   0                   0               0   
3                   0                   0               0   
4                   0                   0               0   

   education_ Assoc-acdm  education_ Assoc-voc  education_ Bachelors  \
0                      0                     0                     1   
1                      0                     0                     1   
2                      0                     0                     0   
3                      0                     0                     0   
4                      0                     0                     1   

   education_ Doctorate  education_ HS-grad  education_ Masters  \
0                     0                   0                   0   
1                     0                   0                   0   
2                     0                   1                   0   
3                     0                   0                   0   
4                     0                   0                   0   

   education_ Preschool  education_ Prof-school  education_ Some-college  \
0                     0                       0                        0   
1                     0                       0                        0   
2                     0                       0                        0   
3                     0                       0                        0   
4                     0                       0                        0   

   marital_status_ Divorced  marital_status_ Married-AF-spouse  \
0                         0                                  0   
1                         0                                  0   
2                         1                                  0   
3                         0                                  0   
4                         0                                  0   

   marital_status_ Married-civ-spouse  marital_status_ Married-spouse-absent  \
0                                   0                   

In [13]:
df_transformed = df[df_num].join(df_catd)
df_transformed.head()

age  fnlwgt  education_num  sex  capital_gain  capital_loss  \
0   39   77516             13    1          2174             0   
1   50   83311             13    1             0             0   
2   38  215646              9    1             0             0   
3   53  234721              7    1             0             0   
4   28  338409             13    0             0             0   

   hours_per_week  income  workclass_ Federal-gov  workclass_ Local-gov  \
0              40       0                       0                     0   
1              13       0                       0                     0   
2              40       0                       0                     0   
3              40       0                       0                     0   
4              40       0                       0                     0   

   workclass_ Never-worked  workclass_ Private  workclass_ Self-emp-inc  \
0                        0                   0                        0   
1                        0                   0                        0   
2                        0                   1                        0   
3                        0                   1                        0   
4                        0                   1                        0   

   workclass_ Self-emp-not-inc  workclass_ State-gov  workclass_ Without-pay  \
0                            0                     1                       0   
1                            1                     0                       0   
2                            0                     0                       0   
3                            0                     0                       0   
4                            0                     0                       0   

   workclass_Unknown  education_ 10th  education_ 11th  education_ 12th  \
0                  0                0                0                0   
1                  0                0                0                0   
2                  0                0                0                0   
3                  0                0                1                0   
4                  0                0                0                0   

   education_ 1st-4th  education_ 5th-6th  education_ 7th-8th  education_ 9th  \
0                   0                   0                   0               0   
1                   0                   0                   0               0   
2                   0                   0                   0               0   
3                   0                   0                   0               0   
4                   0                   0                   0               0   

   education_ Assoc-acdm  education_ Assoc-voc  education_ Bachelors  \
0                      0                     0                     1   
1                      0                     0                     1   
2                      0                     0                     0   
3                      0                     0                     0   
4                      0                     0                     1   

   education_ Doctorate  education_ HS-grad  education_ Masters  \
0                     0                   0                   0   
1                     0                   0                   0   
2                     0                   1                   0   
3                     0                   0                   0   
4                     0                   0                   0   

   education_ Preschool  education_ Prof-school  education_ Some-college  \
0                     0                       0                        0   
1                     0                       0                        0   
2                     0                       0                        0   
3                     0                       0                        0   
4                     0                       0                      

In [23]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

df_transformed[df_num]= scaler.fit_transform(df_transformed[df_num])
df_transformed.head()

/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/data.py:323: DataConversionWarning: Data with input dtype int64 were all converted to float64 by MinMaxScaler.
  return self.partial_fit(X, y)


age    fnlwgt  education_num  sex  capital_gain  capital_loss  \
0  0.301370  0.044302       0.800000  1.0       0.02174           0.0   
1  0.452055  0.048238       0.800000  1.0       0.00000           0.0   
2  0.287671  0.138113       0.533333  1.0       0.00000           0.0   
3  0.493151  0.151068       0.400000  1.0       0.00000           0.0   
4  0.150685  0.221488       0.800000  0.0       0.00000           0.0   

   hours_per_week  income  workclass_ Federal-gov  workclass_ Local-gov  \
0        0.397959     0.0                       0                     0   
1        0.122449     0.0                       0                     0   
2        0.397959     0.0                       0                     0   
3        0.397959     0.0                       0                     0   
4        0.397959     0.0                       0                     0   

   workclass_ Never-worked  workclass_ Private  workclass_ Self-emp-inc  \
0                        0                   0                        0   
1                        0                   0                        0   
2                        0                   1                        0   
3                        0                   1                        0   
4                        0                   1                        0   

   workclass_ Self-emp-not-inc  workclass_ State-gov  workclass_ Without-pay  \
0                            0                     1                       0   
1                            1                     0                       0   
2                            0                     0                       0   
3                            0                     0                       0   
4                            0                     0                       0   

   workclass_Unknown  education_ 10th  education_ 11th  education_ 12th  \
0                  0                0                0                0   
1                  0                0                0                0   
2                  0                0                0                0   
3                  0                0                1                0   
4                  0                0                0                0   

   education_ 1st-4th  education_ 5th-6th  education_ 7th-8th  education_ 9th  \
0                   0                   0                   0               0   
1                   0                   0                   0               0   
2                   0                   0                   0               0   
3                   0                   0                   0               0   
4                   0                   0                   0               0   

   education_ Assoc-acdm  education_ Assoc-voc  education_ Bachelors  \
0                      0                     0                     1   
1                      0                     0                     1   
2                      0                     0                     0   
3                      0                     0                     0   
4                      0                     0                     1   

   education_ Doctorate  education_ HS-grad  education_ Masters  \
0                     0                   0                   0   
1                     0                   0                   0   
2                     0                   1                   0   
3                     0                   0                   0   
4                     0                   0                   0   

   education_ Preschool  education_ Prof-school  education_ Some-college  \
0                     0                       0                        0   
1                     0                       0                        0   
2                     0                       0                        0   
3                     0                       0                        0   
4                     0         

## Part 2 - Fit and present a Logistic Regression

Your data should now be in a state to fit a logistic regression. Use scikit-learn, define your `X` (independent variable) and `y`, and fit a model.

Then, present results - display coefficients in as interpretible a way as you can (hint - scaling the numeric features will help, as it will at least make coefficients more comparable to each other). If you find it helpful for interpretation, you can also generate predictions for cases (like our 5 year old rich kid on the Titanic) or make visualizations - but the goal is your exploration to be able to answer the question, not any particular plot (i.e. don't worry about polishing it).

It is *optional* to use `train_test_split` or validate your model more generally - that is not the core focus for this week. So, it is suggested you focus on fitting a model first, and if you have time at the end you can do further validation.

In [25]:
from sklearn.linear_model import LogisticRegression

y = df_transformed.income
X = df_transformed.drop(columns=['income'])

log_reg = LogisticRegression().fit(X, y)
log_reg.score(X, y)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


0.8522158410368232

In [66]:
cols = df_transformed.columns.drop('income').tolist()
coefs = log_reg.coef_.tolist()[0]
d = {}
for i in range(len(cols)):
  d[cols[i]] = coefs[i]
d

{'age': 1.8144847986650423,
 'capital_gain': 18.005267946508702,
 'capital_loss': 2.558766335089848,
 'education_ 10th': -0.7620609094441765,
 'education_ 11th': -0.7514399620141365,
 'education_ 12th': -0.4548498964821205,
 'education_ 1st-4th': -0.7508381790130364,
 'education_ 5th-6th': -0.6455040166758066,
 'education_ 7th-8th': -0.9974074214722333,
 'education_ 9th': -0.865915021611655,
 'education_ Assoc-acdm': -0.06067068817536697,
 'education_ Assoc-voc': 0.06398952361145097,
 'education_ Bachelors': 0.42794659550092085,
 'education_ Doctorate': 1.1190234147536882,
 'education_ HS-grad': -0.27534429770625174,
 'education_ Masters': 0.679647340323109,
 'education_ Preschool': -1.471798455602201,
 'education_ Prof-school': 1.0421481521353353,
 'education_ Some-college': -0.03820413833596454,
 'education_num': 1.5181002535346944,
 'fnlwgt': 0.9514489283748508,
 'hours_per_week': 2.762182632245099,
 'marital_status_ Divorced': -1.0570479172003284,
 'marital_status_ Married-AF-spous

## Part 3 - Analysis, Interpretation, and Questions

### Based on your above model, answer the following questions

1. What are 3 features positively correlated with income above 50k?
2. What are 3 features negatively correlated with income above 50k?
3. Overall, how well does the model explain the data and what insights do you derive from it?

*These answers count* - that is, make sure to spend some time on them, connecting to your analysis above. There is no single right answer, but as long as you support your reasoning with evidence you are on the right track.

Note - scikit-learn logistic regression does *not* automatically perform a hypothesis test on coefficients. That is OK - if you scale the data they are more comparable in weight.

### Match the following situation descriptions with the model most appropriate to addressing them

In addition to logistic regression, a number of other approaches were covered this week. Pair them with the situations they are most appropriate for, and briefly explain why.

Situations:
1. You are given data on academic performance of primary school students, and asked to fit a model to help predict "at-risk" students who are likely to receive the bottom tier of grades.
2. You are studying tech companies and their patterns in releasing new products, and would like to be able to model and predict when a new product is likely to be launched.
3. You are working on modeling expected plant size and yield with a laboratory that is able to capture fantastically detailed physical data about plants, but only of a few dozen plants at a time.

Approaches:
1. Ridge Regression
2. Quantile Regression
3. Survival Analysis

    What are 3 features positively correlated with income above 50k?
    What are 3 features negatively correlated with income above 50k?
    Overall, how well does the model explain the data and what insights do you derive from it?

3 features positively correlated with income above 50k are capital_gain, capital_loss, and hours_per_week.
3 features negatively correlated with income above 50k are workclass_ Local-gov, occupation_Unknown, and every race.

The model is able to explain about 85% of the variance in income. A lot of the findings I already knew, such as minorities and females generally making less money, but I was surprised that every race, including White, is negatively correlated with income. Being white is just less negatively correlated than some of the other races. I'm not sure how it's possible for every class within a category to be negatively correlated with the target variable.

Situations:

    1. You are given data on academic performance of primary school students, and asked to fit a model to help predict "at-risk" students who are likely to receive the bottom tier of grades.
    2. You are studying tech companies and their patterns in releasing new products, and would like to be able to model and predict when a new product is likely to be launched.
    3. You are working on modeling expected plant size and yield with a laboratory that is able to capture fantastically detailed physical data about plants, but only of a few dozen plants at a time.

Approaches:

    Ridge Regression
    Quantile Regression
    Survival Analysis

The first situation calls for quantile regression. You can determine what percentile performance is the cut-off for "at-risk" and set that percentile as the quantile in the quantile regression. Since you're trying to fit the model to only a subset of the population and not the entire population, quantile regression works great for this.

The third (3rd) situation calls for ridge reression. When there are many more features than observations, a model is at risk of overfitting. When you can't get more observations, the next best thing is to use ridge regression to minimize overfitting and allow the data to be more generalizable.

That leaves survival analysis for the second (2nd) situation. You use survival analysis to predict the duration of time between one event ("birth") and another ("death").